tensorflow/python/compiler/tensorrt/trt_convert.py

```diff
@@ -1133,8 +1143,50 @@ def build(self, input_fn):
       raise RuntimeError("input_fn is None. Method build() needs input_fn "
                          "to be specified in order to build TensorRT engines")
 
+    def _rebuild_func():
+      # Rebuild function from graph_def.
+      reset_converted_func = wrap_function.function_from_graph_def(
+          self._converted_graph_def,
+          [tensor.name for tensor in self._converted_func.inputs],
+          [tensor.name for tensor in self._converted_func.outputs])
+      reset_converted_func.graph.structured_outputs = nest.pack_sequence_as(
+          self._converted_func.graph.structured_outputs,
+          reset_converted_func.graph.structured_outputs)
+      self._converted_func = reset_converted_func
+
+    def _set_profile_generation_mode(value, node):
+      node.attr["_profile_generation_mode"].b = value
+
+    if self._need_trt_profiles:
+      # Enable profile generation.
+      self._for_each_trt_node(self._converted_graph_def,
+                              partial(_set_profile_generation_mode, True))
+      # Profile generation is enabled using the _profile_generation_mode
+      # attribute of the TRTEngineOps. We need to rebuild the function to
+      # change this attribute.
+      _rebuild_func()
+
+    # Use the first input in explicit batch mode to build TensorRT engines
+    # after generating all the profiles. The first input is used but any of
+    # the inputs can be used because the shape of this input does not
+    # determine the engine and instead the shapes collected in profiles
+    # determine the engine.
+    first_input = None
+    # Run inference:
+    #   Builds TRT engines if self._need_trt_profiles is False.
+    #   Builds TRT optimization profiles if self._need_trt_profiles is True.
     for inp in input_fn():
+      if not first_input:
+        first_input = inp
       self._converted_func(*map(ops.convert_to_tensor, inp))
```

Comment:
I don't understand why we need to run the for loop with _converted_func when if self._need_trt_profiles == true. Can you explain?